src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/DiagnosticsClient.cs

```diff
@@ -308,8 +308,15 @@ static IEnumerable<int> GetAllPublishedProcesses()
                     yield return processId;
                 }
             }
-
-            return GetAllPublishedProcesses().Distinct();
+            try 
+            {
+                string[] files = Directory.GetFiles(PidIpcEndpoint.IpcRootPath);
+                return GetAllPublishedProcesses(files).Distinct();
+            }
+            catch ( System.UnauthorizedAccessException ex)
+            {
+                Console.Error.WriteLine("The user that ran this tool does not have access to the NamedPipes pseudo-filesystem. As a result, this tool will not work.");
```

Comment:
I wouldn't suggest having the library write directly to Console.Error(). This library could be used by different apps for different purposes and we don't know what kind of UI each app will have or how it wants to message the errors. What I suggest instead is to create a new Exception type derived from [DiagnosticsClientException](https://github.com/dotnet/diagnostics/blob/main/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/DiagnosticsClientExceptions.cs#L8) and then you can do something like this:

```C#
try
{
    string[] files = Directory.GetFiles(PidIpcEndpoint.IpcRootPath);
    return GetAllPublishedProcesses(files).Distinct();
}
catch( UnauthorizedAccessException ex)
{
    if(PidIpcEndpoint.IpcRootPath.StartsWith(@"\.\\pipe"))
    {
        throw new NamedPipeEnumerationUnauthorizedException($"Enumerating {PidIpcEndpoint.IpcRootPath} is not authorized", ex);
    }
    else
    {
        // We could add other specialized exceptions for other IPCs later if needed, the UnauthorizedException
        // is our general purpose fallback for now.
        throw;
    }
}
``` 

Then in the original code for ps command where you were previously catching the UnauthorizedException you can use `catch (NamedPipeEnumerationUnauthorizedException ex)` instead. Catching the more specific exception type ensures that the error is coming from this one case and not anywhere else. Sorry my suggestion from earlier may have given the wrong impression.